### PR TITLE
[C] Disallow declarations where a statement is required

### LINKED ISSUE #92775: Clang accepts declaration in a statement other than a compound statement
The following code is grammatically incorrect:
```
void func() {
  if (1)
    int x;
}
```
because an iteratation statement must be followed by a secondary-block, which is a kind of statement, which does not accept declarations accept as allowed by compound-statement. GCC properly rejects this code as did Clang 17, so this is a recent regression: https://godbolt.org/z/sj3de1T74

### PR DESCRIPTION
**Pull request by AaronBallman**
This fixes a regression introduced in 8bd06d5b65845e5e01dd899a2deb773580460b89 where Clang began to accept a declaration where a statement is required. e.g.,
```
if (1)
  int x; // Previously accepted, now properly rejected
```

Fixes #92775

### CODE REVIEW

**Review by erichkeane — COMMENTED** [review]
The original patch changed some of the ParsedStmtContext's in ParseOMP that might need to be changed back?  Looking back, I think the intent there was in part to not specify OMP standalone directives, so that part should be reverted as well perhaps?

**Comment by AaronBallman** [comment]
I'm less certain, but CC @jdoerfert @mikerice1969 for more opinions.

The original changes actually fixed a bug in OpenMP related to standalone directives.

> Note, this seems to have fixed an issue with some OpenMP stand-alone
directives not being properly diagnosed as per:
https://www.openmp.org/spec-html/5.1/openmpsu19.html#x34-330002.1.3
(The same requirement exists in OpenMP 5.2 as well.)

File: clang/include/clang/Parse/Parser.h [hunk 1]
```diff
@@ -467,15 +467,18 @@ class Parser : public CodeCompletionHandler {
 
   /// Flags describing a context in which we're parsing a statement.
   enum class ParsedStmtContext {
+    /// This context permits declarations in language modes where declarations
+    /// are not statements.
+    AllowDeclarationsInC = 0x1,
```
Comment (shafik) [inline clang/include/clang/Parse/Parser.h:472]: This feels a little mysterious, I only see that we check it but it is not obvious how it gets set.

**Comment by AaronBallman** [comment]
It gets set a few lines down for `Compound`:

`Compound = AllowDeclarationsInC | AllowStandaloneOpenMPDirectives`

File: clang/include/clang/Parse/Parser.h (hunk 1 above)
Comment (cor3ntin) [reply clang/include/clang/Parse/Parser.h:472]: @shafik L481 below

File: clang/lib/Parse/ParseStmt.cpp [hunk 2]
```diff
@@ -239,7 +239,10 @@ StmtResult Parser::ParseStatementOrDeclarationAfterAttributes(
     auto IsStmtAttr = [](ParsedAttr &Attr) { return Attr.isStmtAttr(); };
     bool AllAttrsAreStmtAttrs = llvm::all_of(CXX11Attrs, IsStmtAttr) &&
                                 llvm::all_of(GNUAttrs, IsStmtAttr);
-    if (((GNUAttributeLoc.isValid() && !(HaveAttrs && AllAttrsAreStmtAttrs)) ||
+    if ((getLangOpts().CPlusPlus || getLangOpts().MicrosoftExt ||
+         (StmtCtx & ParsedStmtContext::AllowDeclarationsInC) !=
+             ParsedStmtContext()) &&
+        ((GNUAttributeLoc.isValid() && !(HaveAttrs && AllAttrsAreStmtAttrs)) ||
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseStmt.cpp:245]: Can you add a standard quote?

**Review by cor3ntin — APPROVED** [review]
LGTM modulo comment

File: clang/lib/Parse/ParseStmt.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/Parse/ParseStmt.cpp:250]: I've added one, let me know if you're happy with it.

Comment (cor3ntin) [reply clang/lib/Parse/ParseStmt.cpp:250]: Thanks (somehow i did not get notified)

**Pull request merged**